nav2_behavior_tree/include/nav2_behavior_tree/bt_cancel_action_node.hpp

```diff
@@ -0,0 +1,169 @@
+
+#include <memory>
+#include <string>
+#include <chrono>
+
+#include "behaviortree_cpp_v3/action_node.h"
+#include "nav2_util/node_utils.hpp"
+#include "rclcpp_action/rclcpp_action.hpp"
+#include "nav2_behavior_tree/bt_conversions.hpp"
+
+using namespace std::literals;
+
+namespace nav2_behavior_tree
+{
+
+/**
+ * @brief Abstract class representing an action for cancelling BT node
+ * @tparam ActionT Type of action
+ */
+template<class ActionT>
+class BtCancelActionNode : public BT::ActionNodeBase
+{
+public:
+  /**
+   * @brief A nav2_behavior_tree::BtCancelActionNode constructor
+   * @param xml_tag_name Name for the XML tag for this node
+   * @param action_name Action name this node creates a client for
+   * @param conf BT node configuration
+   */
+  BtCancelActionNode(
+    const std::string & xml_tag_name,
+    const std::string & action_name,
+    const BT::NodeConfiguration & conf)
+  : BT::ActionNodeBase(xml_tag_name, conf), action_name_(action_name)
+  {
+    node_ = config().blackboard->template get<rclcpp::Node::SharedPtr>("node");
+    callback_group_ = node_->create_callback_group(
+      rclcpp::CallbackGroupType::MutuallyExclusive,
+      false);
+    callback_group_executor_.add_callback_group(callback_group_, node_->get_node_base_interface());
+
+    // Get the required items from the blackboard
+    server_timeout_ =
+      config().blackboard->template get<std::chrono::milliseconds>("server_timeout");
+    getInput<std::chrono::milliseconds>("server_timeout", server_timeout_);
+
+    std::string remapped_action_name;
+    if (getInput("server_name", remapped_action_name)) {
+      action_name_ = remapped_action_name;
+    }
+    createActionClient(action_name_);
+
+    // Give the derive class a chance to do any initialization
+    RCLCPP_DEBUG(
+      node_->get_logger(), "\"%s\" BtCancelActionNode initialized",
+      xml_tag_name.c_str());
+  }
+
+  BtCancelActionNode() = delete;
+
+  virtual ~BtCancelActionNode()
+  {
+  }
+
+  /**
+   * @brief Create instance of an action client
+   * @param action_name Action name to create client for
+   */
+  void createActionClient(const std::string & action_name)
+  {
+    // Now that we have the ROS node to use, create the action client for this BT action
+    action_client_ = rclcpp_action::create_client<ActionT>(node_, action_name, callback_group_);
+
+    // Make sure the server is actually there before continuing
+    RCLCPP_DEBUG(node_->get_logger(), "Waiting for \"%s\" action server", action_name.c_str());
+    action_client_->wait_for_action_server();
+  }
+
+  /**
+   * @brief Any subclass of BtCancelActionNode that accepts parameters must provide a
+   * providedPorts method and call providedBasicPorts in it.
+   * @param addition Additional ports to add to BT port list
+   * @return BT::PortsList Containing basic ports along with node-specific ports
+   */
+  static BT::PortsList providedBasicPorts(BT::PortsList addition)
```

Comment:
https://github.com/ros-planning/navigation2/blob/63feaf70780d1d41a694d6b53aadc6005cdb7ce2/nav2_behavior_tree/include/nav2_behavior_tree/plugins/action/back_up_action.hpp#L52-L61

Do something in this style rather than the 2 functions, its a little expanded and not in the style of the other ones

Reply:
Then you don't need to implement one of these functions, I don't believe

Reply:
Unfortunately, that will not be possible because, `back_up_action` was derived from the `bt_action_node`. Also, on top of the `providedBasicPorts`, we can find a comment stated as follows

https://github.com/ros-planning/navigation2/blob/63feaf70780d1d41a694d6b53aadc6005cdb7ce2/nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp#L96-L101